src/components/DBEntryManagerPanel/AddOrEditDBEntry/index.tsx

```diff
@@ -0,0 +1,614 @@
+/*------------------------------------------------------------------------*/
+
+const AddorEditDBEntry: React.FC<Props> = (props) => {
+  /*------------------------------------------------------------------------*/
+  /* -------------------------------- Setup ------------------------------- */
+  /*------------------------------------------------------------------------*/
+
+  /* -------------- Props ------------- */
+
+  // Destructure all props
+  const {
+    onFinished,
+    entryFields,
+    DBEntryToEdit,
+    validationFunction,
+    objectModifier,
+    idPropName,
+    endpoint,
+    entries,
+  } = props;
+
+  /* -------------- State ------------- */
+
+  // Initial state
+  const initialState: State = {
+    entry: DBEntryToEdit || {},
+    saving: false,
+  };
+
+  // Initialize state
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  // Destructure common state
+  const {
+    entry,
+    saving,
+  } = state;
+
+  /*------------------------------------------------------------------------*/
+  /* ------------------------- Component Functions ------------------------ */
+  /*------------------------------------------------------------------------*/
+
+  /**
+   * Save changes to the DBEntry and then finish
+   * @author Yuen Ler Chow
+   */
+  const save = async () => {
+    // Start the save loading indicator
+    dispatch({ type: ActionType.StartSave });
+
+    // add all default values to the entry
+    entryFields.forEach((field) => {
+      if (field.defaultValue && !entry[field.objectKey]) {
+        entry[field.objectKey] = field.defaultValue;
+      }
+    });
+
+    const modifiedEntry = objectModifier ? objectModifier(entry) : entry;
+
+    // Send to server
+    try {
+      await visitServerEndpoint({
+        path: endpoint,
+        method: 'POST',
+        params: {
+          item: JSON.stringify(modifiedEntry),
+        },
+      });
+
+      // Finish
+      onFinished(entry);
+    } catch (err) {
+      return showFatalError(err);
+    }
+  };
+
+  /**
+   * Cancel and return without saving
+   * @author Gabe Abrams
+   */
+  const cancel = async () => {
+    onFinished(undefined);
+  };
+
+  /*------------------------------------------------------------------------*/
+  /* ------------------------------- Render ------------------------------- */
+  /*------------------------------------------------------------------------*/
+
+  /*----------------------------------------*/
+  /* ---------------- Views --------------- */
+  /*----------------------------------------*/
+
+  // Body
+  let body: React.ReactNode;
+
+  /* ------------- Loading ------------ */
+
+  if (saving) {
+    body = (
+      <LoadingSpinner />
+    );
+  }
+
+  /* -------------- Form -------------- */
+
+  if (!saving) {
+    // create validation boolean array for each field
+    let validationError = '';
+
+    for (let i = 0; i < entryFields.length; i += 1) {
```

Comment:
This could probably be a `forEach`

Reply:
is there a way to break out of foreach early